xdsl/ir.py

```diff
@@ -632,9 +640,9 @@ def __hash__(self) -> int:
 
     @classmethod
     @property
-    def irdl_definition(cls) -> OpDef:
+    def irdl_definition(cls) -> OpDef | None:
         """Get the IRDL operation definition."""
-        ...
+        return None
```

Comment:
I don't quite like this. AFAIU, Operation is a abstract class and this is how abstract classes should be implemented in Python. Can you add the `abstract_method` decorator, maybe this solves the problem?

Reply:
To be honest there's a different thing that's the root cause of the ugliness, which is a break in the layering of `irdl` being on top of `ir`. This Operation class should not know anything about irdl, but does, it feels that we should decide whether all users of `Operation` would want to implement this function, and thus import `irdl.py` as well, or whether a default implementation that is safe is ok

Reply:
Oh yes, you are right!
IMO, we should try find a different design that does not have this circular problem. Maybe the the dialect objects actually become useful finally? @math-fehr 

O/w the default implementation is fine as well...

Reply:
My suggestion would be to have an `IRDLOperation` that is also abstract and that `@irdl_op_definition` targets

Reply:
My gut feeling is that this would be an OK thing to defer until we're happy with extensive breaking changes. `Operation` could also be renamed to `AbstractOperation` to make things more explicit

Reply:
I'm actually fine using `IRDLOperation`, but I do find ugly that we would have to both write `IRDLOperation` and `irdl_op_definition`.
One way of doing this could be to use metaclasses rather than decorators (they are equivalent), but then we would need to do that for attributes as well.
I do not think this is a significant problem now though, so I would just not worry about it (At least, I feel there are other more important problems).

Reply:
I mean, to be honest, I would prefer going the route of "everything uses IRDL" rather than the root of "We have a lot of support for anyone not wanting to use IRDL, even though no one is currently doing that"

Reply:
That works for me, although the issue of circular dependencies still exists. Even if the only supported way to use it is IRDL, it would still make sense for `ir.py` not to depend on `irdl.py`